doc/design/parallel-node-config.md

```diff
@@ -0,0 +1,214 @@
+
+Config Daemon will read SriovNetworkNodeState annotation `sriovnetwork.openshift.io/desired-state` and write both
+node annotation, `sriovnetwork.openshift.io/state` and SriovNetworkNodeState annotation `sriovnetwork.openshift.io/desired-state`.
+
+*NOTE:* In the future we are going to drop the node annotation and only use the SriovNetworkNodeState
+
+Draining procedure:
+
+1. config daemon mark the node as `Drain_Required` or `Reboot_Required`
+2. operator drain controller reconcile loop find the node SriovNetworkPoolConfig
+   1. if number of `Draining` nodes is great or equal to the `MaxUnavailable` the operator will re-queue the request
+   2. if number of `Draining` nodes is lower than the `MaxUnavailable` the operator will start the draining process
+5. on Openshift platform we will pause the machine config pool related to the node
+6. the operator will start the drain process
+   1. if `Drain_Required` the operator will remove ONLY pods used sriov devices
+   2. if `Reboot_Required` the operator will remove ALL the pods on the system
+9. operator moves the `sriovnetwork.openshift.io/current-state` annotation to `DrainComplete`
+10. daemon will continue to the configuration when it's done it will move back the `sriovnetwork.openshift.io/state` annotation to `Idle`
+11. operator runs the complete drain to remove the cordon and mark the `sriovnetwork.openshift.io/current-state` annotation to `Idle`
+
+### API Extensions
+
+#### Extend existing CR SriovNetworkPoolConfig
+SriovNetworkPoolConfig is used only for OpenShift to provide configuration for
+OVS Hardware Offloading. We can extend it to add configuration for the drain
+pool. E.g.:
+
+```golang
+// SriovNetworkPoolConfigSpec defines the desired state of SriovNetworkPoolConfig
+type SriovNetworkPoolConfigSpec struct {
+    ...
+	
+    // nodeSelector specifies a label selector for Nodes
+    NodeSelector *metav1.LabelSelector `json:"nodeSelector,omitempty"`
+
+    // maxUnavailable defines either an integer number or percentage
+    // of nodes in the pool that can go Unavailable during an update.
+    //
+    // A value larger than 1 will mean multiple nodes going unavailable during
+    // the update, which may affect your workload stress on the remaining nodes.
+    // Drain will respect Pod Disruption Budgets (PDBs) such as etcd quorum guards,
+    // even if maxUnavailable is greater than one.
+    MaxUnavailable *intstr.IntOrString `json:"maxUnavailable,omitempty"`
+}
+```
+
+```yaml
+apiVersion: v1
+kind: SriovNetworkPoolConfig
+metadata:
+  name: pool-1
+  namespace: network-operator
+spec:
+  maxUnavailable: "20%"
+  nodeSelectorTerms:
+    - matchExpressions:
+      - key: some-label
+        operator: In
+        values:
+          - val-2
+    - matchExpressions:
+      - key: other-label
+        operator: "Exists"
+```
+
+Default pool should be always created with the lowest priority:
```

Comment:
also i thought we want to get rid of "default" objects created by the operator.

Reply:
I don't know if we want to drop the default for a pool.
This is to explicitly say `maxUnavailable=1`

also in the future, I would like to add an annotation or field to the NodeState to say what pool the node is part of
and in the pool how many nodes are part of a pool and thinks like that

Reply:
im fine with requiring an explicit pool but i dont want the operator to be the one creating it (which means we need to delete it at some point)

as we do try to avoid creating default objects (e.g sriov policy and operator config)

Reply:
ok I am fine not having the default pool created let me update the doc